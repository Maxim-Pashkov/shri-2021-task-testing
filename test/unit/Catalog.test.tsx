import { it, expect, describe } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import { getApplication } from './getApplication';
import { Application } from '../../src/client/Application';

describe('Catalog test', () => {
    it('Catalog page loading is visible', () => {
        const { application, history } = getApplication(
            () => <Application />,
            {}
        );
        history.push('/catalog');

        const { queryByText } = render(application);

        expect(queryByText(/loading/i)).toBeTruthy();
    });

    it('Catalog page is visible', async () => {
        const { application, history, store } = getApplication(
            () => <Application />,
            {}
        );
        history.push('/catalog');

        const { queryByText } = render(application);

        await new Promise((resolve) => store.subscribe(() => resolve(true)));

        expect(queryByText(/loading/i)).toBeFalsy();
    });

    it('Product are displayed in Catalog', async () => {
        const { application, history, store } = getApplication(
            () => <Application />,
            {}
        );
        history.push('/catalog');

        const { queryAllByTestId } = render(application);

        await new Promise((resolve) => store.subscribe(() => resolve(true)));

        const ids = store.getState().products.map((product) => product.id);
        const elements = ids.map(
            (id) => +queryAllByTestId(id)[0]?.getAttribute('data-testid')
        );

        expect(ids).toEqual(expect.arrayContaining(elements));
    });
});
