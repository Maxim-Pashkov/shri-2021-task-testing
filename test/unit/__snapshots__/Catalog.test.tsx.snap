// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog test Catalog page is visible 1`] = `
<div
  class="Catalog"
>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <h1>
        Catalog
      </h1>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-12 col-sm-6 col-md-4 col-lg-3"
      data-testid="2"
    >
      <div
        class="ProductItem card w-100 mb-4"
        data-testid="2"
      >
        <img
          class="Image card-img-top"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
        />
        <div
          class="card-body"
        >
          <h5
            class="ProductItem-Name card-title"
          >
            test name 2
          </h5>
          <p
            class="ProductItem-Price card-text"
          >
            $
            10
          </p>
          <a
            class="ProductItem-DetailsLink card-link"
            href="/catalog/2"
          >
            Details
          </a>
        </div>
      </div>
    </div>
    <div
      class="col-12 col-sm-6 col-md-4 col-lg-3"
      data-testid="3"
    >
      <div
        class="ProductItem card w-100 mb-4"
        data-testid="3"
      >
        <img
          class="Image card-img-top"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
        />
        <div
          class="card-body"
        >
          <h5
            class="ProductItem-Name card-title"
          >
            test name 3
          </h5>
          <p
            class="ProductItem-Price card-text"
          >
            $
            11
          </p>
          <a
            class="ProductItem-DetailsLink card-link"
            href="/catalog/3"
          >
            Details
          </a>
        </div>
      </div>
    </div>
    <div
      class="col-12 col-sm-6 col-md-4 col-lg-3"
      data-testid="4"
    >
      <div
        class="ProductItem card w-100 mb-4"
        data-testid="4"
      >
        <img
          class="Image card-img-top"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
        />
        <div
          class="card-body"
        >
          <h5
            class="ProductItem-Name card-title"
          >
            test name 4
          </h5>
          <p
            class="ProductItem-Price card-text"
          >
            $
            12
          </p>
          <a
            class="ProductItem-DetailsLink card-link"
            href="/catalog/4"
          >
            Details
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Catalog test Catalog page loading is visible 1`] = `
<div
  class="Catalog"
>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <h1>
        Catalog
      </h1>
    </div>
  </div>
  <div
    class="row"
  >
    LOADING
  </div>
</div>
`;
